Feature/autonomy

thomas-gleiforst opened the pull request:
This pull request of the Feature/autonomy branch contains everything needed for the autonomy tab of RAM. This includes:
- a state diagram
- a log that shows waypoints sent to autonomy, logs from autonomy, and flashes green when an AR tag has been reached
- controls to start and stop autonomy as well as send and clear waypoints
- display obstacle avoidance, ar tag detection, or whatever other feeds are being streamed by autonomy

thomas-gleiforst committed:
Add shell for autonomy view

thomas-gleiforst committed:
Add controls view

thomas-gleiforst committed:
Add activity framework

note: updates the manifest to move current log from error to telemetry

thomas-gleiforst committed:
Added mediocre state diagram

thomas-gleiforst committed:
Small stylistic improvements

thomas-gleiforst committed:
Fix name of alternate windows

thomas-gleiforst committed:
Log and flash colors when reaching a marker

thomas-gleiforst committed:
Update camera mappings since autonomy downgraded motion

thomas-gleiforst committed:
Remove testing "reached: button

EliVerbrugge commented on Electron/src/Core/components/Cameras.tsx:
```diff
@@ -47,12 +47,7 @@ class Cameras extends Component<IProps, IState> {
-        "http://192.168.1.139:8081",
-        "http://192.168.1.139:8082",
-      ],
+      baseAddress: ["http://192.168.1.50:8080", "http://192.168.1.51:8080", "http://192.168.1.139:8080"],
```
Could you parse these from the manifest? Idk, might be more clear.

EliVerbrugge commented on Electron/src/RAM/Autonomy/components/Activity.tsx:
```diff
@@ -0,0 +1,84 @@
+
+  Log(data: string): void {
+    let text = this.state.ActivityText
+    text += `${data} \n` // set this to be that variable and time and newline
```
If possible, a simple timestamp might be nice (especially if we leave Base Station running for a while)

EliVerbrugge reviewed (commented):
Ty, ty. Looks great.

thomas-gleiforst committed:
Pull camera IPs from manifest

Note: need to change ips on camera pis before this is merged into dev

thomas-gleiforst committed:
Add timestamp to activity and log

This was done previously in Ryan's Feature/Consol [sic] branch, which has since been redone and deleted, but this feature was missed

Co-Authored-By: rkbhkp <54925741+rkbhkp@users.noreply.github.com>

amwgvf commented on Electron/src/Core/components/Cameras.tsx:
```diff
@@ -39,20 +39,15 @@ interface IProps {
     super(props)
     this.state = {
       currentCamera: this.props.defaultCamera,
-      baseAddress: [
```
I'm confused why you made this change, doesn't it do functionally the same thing?

thomas-gleiforst replied:
The change is much more future proof. Grabbing the IPs from the manifest means if the camera IP or autonomy IP changes for any reason, we don't have to change them again in the camera file.

amwgvf commented on Electron/src/RAM/Autonomy/components/Activity.tsx:
```diff
@@ -0,0 +1,84 @@
+    this.setState({ backgroundColor: "green" })
+    this.Log("Reached waypoint!")
+    const reachInterval = setInterval(() => {
+      const backgroundColor: string = this.state.backgroundColor === "green" ? "white" : "green"
```
can we/do we use the background colors in the same way on any other components? i feel like it could be cool

thomas-gleiforst replied:
The intent of this function is to strobe the background several times between green and white, and ensure it always ends on white. This is a requirement of SAR, as when the rover reaches an AR tag and stops it must be indicated both on rover and on basestation. In most cases throughout basestation a direct toggle of background color isn't useful (especially not useful enough to abstract this out into any kind of converter), but there is a [ColorConverter](https://github.com/MissouriMRDT/BaseStation_Software/blob/dev/Electron/src/Core/ColorConverter.tsx) in Core that takes in several parameters to allow responsive background colors on the screen.

amwgvf commented on Electron/src/RAM/Autonomy/components/Activity.tsx:
```diff
@@ -0,0 +1,84 @@
+    setTimeout(() => {
+      clearInterval(reachInterval)
+      this.setState({ backgroundColor: "white" })
+    }, 4000)
```
why did you choose 4000 for the timeout time? and why 250 on line 66?

thomas-gleiforst replied:
The 4000 timeout time here is to only pulse the colors of the screen for 4 seconds, as that's long enough to be noticeable but not so long that its annoying. The other timeout is 250, implying the color should toggle 4 times per second (although it isn't quite that fast since setState is asynchronous and takes a slightly different amount of time each cycle)

amwgvf commented on Electron/src/RAM/Autonomy/components/StateDiagram.tsx:
```diff
@@ -0,0 +1,186 @@
+    rovecomm.on("CurrentState", (data: any) => this.updateStateDiagram(data))
+  }
+
+  componentDidMount() {
```
i'm confused lol. what does this function do?

thomas-gleiforst replied:
componentDidMount is a built in function that gets called whenever the component successfully mounts (or is fully constructed). That ensures this function runs once (and only once) after all variables have been created and defaulted. We call this.updateStateDiagram[-1] to set the clear the state diagram.

amwgvf commented on Electron/src/RAM/Autonomy/components/StateDiagram.tsx:
```diff
@@ -0,0 +1,186 @@
+    context.textAlign = "right"
+    context.fillStyle = "black"
+    context.fillText(text, centerW - 25, 130)
+    context.rect(centerW - 20, 110, 40, 40)
```
this might be dumb but what is this the center of?

thomas-gleiforst replied:
centerW is a constant set on line 66, whose value is the canvas / 2. The canvas should extend to be the size of the entire div. So this is the center of the space that we are trying to draw in.
